auto_dev/protocols/adapters.py

```diff
@@ -103,25 +117,47 @@ class FileAdapter:
     def __getattr__(self, name: str):
         return getattr(self.wrapped, name)
 
+    @property
+    def enum_names(self) -> set[str]:
+        return {m.name for m in self.enums}
+
+    @property
+    def message_names(self) -> set[str]:
+        return {m.name for m in self.messages}
+
     @classmethod
     def from_file(cls, file: File) -> FileAdapter:
         """Convert a `File` into `FileAdapter`, handling messages recursively."""
 
-        elements = {camel_to_snake(t.__name__): [] for t in FileElement.__args__}
-
-        for element in file.file_elements:
+        file_elements = []
+        grouped_elements = {camel_to_snake(t.__name__): [] for t in FileElement.__args__}
+        for i, element in enumerate(file.file_elements):
             key = camel_to_snake(element.__class__.__name__)
-            elements[key].append(element)
+            if isinstance(element, Message):
+                element = MessageAdapter.from_message(element)
+            file_elements.append(element)
+            grouped_elements[key].append(element)
 
-        return cls(
+        file_adapter = cls(
             wrapped=file,
+            file_elements=file_elements,
             syntax=file.syntax,
-            imports=elements["import"],
-            packages=elements["package"],
-            options=elements["option"],
-            messages=[MessageAdapter.from_message(m) for m in elements["message"]],
-            enums=elements["enum"],
-            extensions=elements["extension"],
-            services=elements["service"],
-            comments=elements["comment"]
+            imports=grouped_elements["import"],
+            packages=grouped_elements["package"],
+            options=grouped_elements["option"],
+            messages=grouped_elements["message"],
+            enums=grouped_elements["enum"],
+            extensions=grouped_elements["extension"],
+            services=grouped_elements["service"],
+            comments=grouped_elements["comment"]
         )
+
+        def set_file_adapter(message: MessageAdapter):
+            message.file = file_adapter
+            for nested_message in message.messages:
+                set_file_adapter(nested_message)
+
+        for message in file_adapter.messages:
+            set_file_adapter(message)
```

Comment:
allow for easy access to check if a particular `Field.type` or `MapField.value_type` is referencing a top level `Enum` or `Message`, information that is required for rendering.